module/zfs/zvol.c

```diff
@@ -108,6 +109,29 @@ static struct hlist_head *zvol_htable;
 
 static struct ida zvol_ida;
 
+typedef struct zvol_aggsum_stats {
+	aggsum_t zvol_as_writes;
+	aggsum_t zvol_as_nwritten;
+	aggsum_t zvol_as_reads;
+	aggsum_t zvol_as_nread;
+} zvol_aggsum_stats_t;
+
+typedef struct zvol_kstats {
+	kstat_named_t zk_writes;
+	kstat_named_t zk_nwritten;
+	kstat_named_t zk_reads;
+	kstat_named_t zk_nread;
+	kstat_named_t zk_ds_name;
+} zvol_kstats_t;
+
+static zvol_kstats_t empty_zvol_kstats = {
+	{ "writes",	KSTAT_DATA_UINT64 },
+	{ "nwritten",	KSTAT_DATA_UINT64 },
+	{ "reads",	KSTAT_DATA_UINT64 },
+	{ "nread",	KSTAT_DATA_UINT64 },
+	{ "dataset_name",	KSTAT_DATA_STRING },
+};
```

Comment:
Can we have a generic "dataset kstats" rather than separate zvol and zpl kstats?  zvol vs zpl would need to call into the generic dataset kstats from different places, but I would think that we could use the same structs, and zvol_kstats_update(), zvol_kstats_create(), zvol_kstats_destroy(), and update funcs could be shared.

You could add a new .c / .h for the dataset kstats, since this doesn't seem closely tied in with e.g. objset code.